src/materialized/src/http.rs

```diff
@@ -181,7 +182,27 @@ impl Server {
                 };
                 coord_client.terminate().await;
                 res
-            }
+            };
+            // Hyper will drop the future if the client goes away, in an effort
+            // to eagerly cancel work. But the design of the coordinator
+            // requires that the future be polled to completion in order for
+            // cleanup to occur. Specifically:
+            //
+            //   * The `SessionClient` *must* call `terminate` before it is
+            //     dropped.
+            //   * A peek receiver must not be dropped before it receives a
+            //     message.
+            //
+            // Not observing this rule leads to invariant violations that panic;
+            // see #6278 for an example.
+            //
+            // The fix here is to wrap the future in a combinator that will call
+            // `tokio::spawn` to poll it to completion if Hyper gives up on it.
+            // A bit weird, but it works, and hides this messiness from the code
+            // in the future itself. If Rust ever supports asynchronous
+            // destructors ("AsyncDrop"), those will admit a more natural
+            // solution to the problem.
+            future.spawn_if_canceled()
```

Comment:
It's funny to me that this is basically what less-sophisticated web frameworks do in scripting languages, too: On a hangup they'll just keep going and then discard the response.

Seems like the right thing to do, though (and allows for a decent mental model of what is going on).

Reply:
Hah, I know, right?